src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,282 @@ impl<F: Field> ConstraintSystem<F> {
+                        exclusion_matrix[i][j] = true;
+                    }
+                }
+            }
+        }
+
+        // Keep track of the combinations; the ith element of this vector
+        // represents the ith combination and contains a vector of (simple)
+        // selectors that are included.
+        let mut combinations = vec![];
+
+        // We will not increase the degree of the constraint system, so we limit
+        // ourselves to the largest existing degree constraint.
+        let max_degree = self.degree();
+
+        // Simple selectors that we've added to combinations already.
+        let mut seen = vec![false; self.num_selectors];
+
+        let mut selector_map = vec![0; self.num_selectors];
+
+        for i in 0..self.num_selectors {
+            if complex_selectors[i] {
+                selector_map[i] = combinations.len();
+                // Complex selectors are on their own.
+                combinations.push(vec![i]);
+
+                // NB: We don't update `seen` here as it's a complex selector
+                // that will be ignored anyway.
+            } else {
+                // This is a simple selector.
+                if seen[i] {
+                    // This simple selector was added to a previous combination.
+                    continue;
+                }
+                selector_map[i] = combinations.len();
+                seen[i] = true; // We're definitely adding this to a combination.
+                let mut combination = vec![i];
+                assert!(degrees[i] <= max_degree); // Otherwise, self.degree() is wrong.
+
+                // Keep track of how large the substitution is.
+                let mut usage = 1;
+
+                // Keep track of how large of a gate any of the members of this
+                // combination can support.
+                let mut room = max_degree - degrees[i];
+
+                // Try to find other selectors that can join this one.
+                for j in (i + 1)..self.num_selectors {
+                    // Is there room for the new selector?
+                    if room == 0 {
+                        // Short circuit here, we've hit our limit.
+                        break;
+                    }
+
+                    // Skip complex selectors.
+                    if complex_selectors[j] {
+                        continue;
+                    }
+
+                    // Skip selectors that have been added to previous combinations
+                    if seen[j] {
+                        continue;
+                    }
+
+                    // Are the two selectors excluded from co-existing?
+                    // j > i and so (j, i) is always on the lower diagonal
+                    if exclusion_matrix[j][i] {
+                        continue;
+                    }
+
+                    // Can the new selector join the combination?
+                    if degrees[j] + usage <= max_degree {
```

Comment:
This has an off-by-one error: usage hasn't yet been incremented, and so the actual maximum degree of any gate involving selector j is `degrees[j] + usage + 1`.

(The `room` check above accounted for the case where the degree bound would be violated by adding another term to a gate *not* involving selector j.)

Reply:
Added some notes about this in various places.